src/procedural-macros.md

```diff
@@ -1,24 +1,288 @@
+
+This crate primarily contains a [`TokenStream`] type. Procedural macros operate
+over *token streams* instead of AST nodes, which is a far more stable interface
+over time for both the compiler and for procedural macros to target. A
+*token stream* is roughly equivalent to `Vec<TokenTree>` where a `TokenTree`
+can roughly be thought of as lexical token. For example `foo` is an `Ident`
+token, `.` is a `Punct` token, and `1.2` is a `Literal` token. The `TokenStream`
+type, unlike `Vec<TokenTree>`, is cheap to clone.
+
+All tokens have an associated `Span`. A `Span` is an opaque value that cannot
+be modified but can be manufactured. `Span`s represent an extent of source
+code within a program and are primarily used for error reporting. You can modify
+the `Span` of any token.
+
+### Procedural macro hygiene
+
+Procedural macros are *unhygienic*. This means they behave as if the output
+token stream was simply written inline to the code it's next to. This means that
+it's affected by external items and also affects external imports.
+
+Macro authors need to be careful to ensure their macros work in as many contexts
+as possible given this limitation. This often includes using absolute paths to
+items in libraries (for example, `::std::option::Option` instead of `Option`) or
+by ensuring that generated functions have names that are unlikely to clash with
+other functions (like `__internal_foo` instead of `foo`).
+
+### Function-like procedural macros
+
+Function-like procedural macros define new invokable macros.
+
+These macros are defined by a [public] [function] with the `proc_maco`
+[attribute] and a signature of `(TokenStream) -> TokenStream`. The input
+[`TokenStream`] is what is inside the delimiters of the macro invocation and the
+output [`TokenStream`] replaces the entire macro invocation. It may contain an
+arbitrary number of [items]. The returned [`TokenStream`] cannot include any
+[macro] definitions.
+
+For example, the following macro definition ignores its input and outputs a
+function `answer` into its scope.
+
+```rust,ignore
+extern crate proc_macro;
+use proc_macro::TokenStream;
+
+#[proc_macro]
+pub fn make_answer(_item: TokenStream) -> TokenStream {
+    "fn answer() -> u32 { 42 }".parse().unwrap()
+}
+```
+
+And then we use it a binary crate to print "42" to standard output.
 
 ```rust,ignore
+extern crate proc_macro_examples;
+use proc_macro_examples::make_answer;
+
+make_answer!();
+
+fn main() {
+    println!("{}", answer());
+}
+```
+
+These macros are only invokable in [modules]. They cannot even be invoked to
+make [item declaration statements]. Furthermore, they must either be invoked
+with curly braces and no semicolon or a different delimiter followed by a
+semicolon. For example, `make_answer` from the previous example can be invoked
+as `make_answer!{}`, `make_answer!();` or `make_answer![];`.
+
+These macros cannot expand to syntax that defines new `macro_rule` style macros.
+
+### Derive mode macros
```

Comment:
What's the point of "mode" in "derive mode"?
It's like a random unrelated word just inserted there.
I've seen the wording "derive mode" in a couple of compiler errors maybe, but that's all.

The phrasing "define new modes for the `derive` attribute" doesn't make sense to me.
Ok, I'm not a native speaker, let's look into [a dictionary](https://en.oxforddictionaries.com/definition/mode) - "A way or manner" / "A fashion or style" - still doesn't make any sense.
In the first approximation, `derive` just applies arbitrary macros passed to it (`#[derive(A, B, C)] ITEM` -> `#[A] #[B] #[C] ITEM`), how are those macros "modes"?

\</rant>

Reply:
The main point, as I see it, is that it lets us talk about the `A`, `B`, `C` inside `#[derive(A, B, C)]` without needing to always call them macros. If we want to talk about `A` without caring that its a macro, and we just call them derive macros, then what do we call it? We could call it a "derive", but that does not jive well. "The derive attribute takes derives" is also weird. So we need some name to distinguish them. My original thought was to go with "deriver", but then I saw said compiler error that calls them a derive mode. And mode makes sense to me.

When I think of modes, I think of distinct ways of operating something. My oven has modes for "bake", "boil", some other one I never use, and "clean". Firefox has safe mode and private browsing mode. Businesses operate in stealth mode or growth mode or whatever fancy mode they want. That said, if you want to get it changed to "deriver" or something, file an issue against rust-lang/rust. I defer terminology to what it uses when I can.

Reply:
>I saw said compiler error that calls them a derive mode.

Yeah, I've seen those too and always thought they were bad, and wanted to get rid of them.
That's why when I'd seen that terminology reused I immediately thought it wasn't a good idea.
<sub><sup>(Please, excuse my past tenses if I got them all wrong.)</sup></sub>

I suspect the terminology came from the time when only a limited set of built-in derives existed.
"Mode" kinda implies that you have several of them and need to choose one, i.e. something `enum`-like.

>without needing to always call them macros

Why? They are macros.
If `#[proc_macro_attribute]` defines an "attribute macro", `#[proc_macro_derive]` defines a "derive macro", then we have a symmetry between language and docs.
"`derive` attribute takes derive macros as arguments" - sounds ok to me.